### PR TITLE
feat(synthetics): add vpc configuration

### LINKED ISSUE #9954: [synthetics] Make it possible to set vpcConfig in synthetics.Canary for vpc private API testing
When creating a Canary using synthetics.Canary there is no way to set the vpc config which is needed for invocations of VPC-private APIs.

If you create the canary using CfnCanary you can configure the VpcConfig in the https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-synthetics.CfnCanaryProps.html

So there is a workaround, but by using CfnCanary you have to create the role, permission, s3 bucket (artifacts) etc. by hand also - so you end up with a lot of cdk code for even simple canaries.

### Use Case

It is not possible to configure api testing of vpc private api calls without this option when using the synthetics.Canary - and using CfnCanary is quite verbose / easy to get wrong in comparison.

### Proposed Solution

Add vpcConfig option to the synthetics.CanaryProps and 

### Other
If it is not an option to add vpc config to the synthetics.Canary, then please provide an example of how to use the vpc config in the CfnCanary.

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by RichiCoder1**
This PR adds vpc support to synthetics and is a continuation of #11865.

See [Running a canary on a vpc](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_VPC.html).

Fixes #9954

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-synthetics/test/integ.canary.expected.json
Comment (RichiCoder1) [inline packages/@aws-cdk/aws-synthetics/test/integ.canary.expected.json:124]: This is a relic of the previous PR, is this behavior that we want? Or should Timeout only be introduced when any other `RunConfig` properties are specified?

File: packages/@aws-cdk/aws-synthetics/lib/schedule.ts
Comment (RichiCoder1) [inline packages/@aws-cdk/aws-synthetics/lib/schedule.ts:73]: This is heckin' hacky, but I couldn't come up with a way around it aside from having `Schedule` store the input durations.

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (RichiCoder1) [inline packages/@aws-cdk/aws-synthetics/lib/canary.ts:460]: This code was basically lifted from `Function`, let me know if it needs refiniment!

**Comment by kaizen3031593** [comment]
Hi @RichiCoder1, thanks for opening this PR! As a first step, lets split this into two PRs, one for runConfig and one for vpcConfig. Specifically for `runConfig`, can you tell me the value add for exposing the rest of those properties over just using escape hatches? I don't see much of a use case for them and we're certainly not blocking anyone. That's why I didn't implement them in the L2 and I don't see anyone knocking down the door for them (the way they were for environment variables). `vpcConfig` might be helpful, but also might be harder to get right.

**Comment by RichiCoder1** [comment]
> Specifically for runConfig, can you tell me the value add for exposing the rest of those properties over just using escape hatches? I don't see much of a use case for them and we're certainly not blocking anyone. That's why I didn't implement them in the L2 and I don't see anyone knocking down the door for them (the way they were for environment variables)

Partially completeness, partially convenience. I'm personally using the `timeout` and `tracing` properties, and while "normal" it's weird to drop the Cfn just to set them when they're simple to implement.

> As a first step, lets split this into two PRs, one for runConfig and one for vpcConfig

I probably should have seen this ask coming, I'll see what I can do :). Before I do so, do we want the VPC configuration options to be top level options, or a sub config like I have in this PR?

**Comment by kaizen3031593** [comment]
I don't think its partially complete -- environment variables is its own thing separate from timeout and tracing. The canary L2 is not simply a shell for the L1 and we do not need to adhere to how cfn clusters properties; this means that supplying env variables in the L2 is visible at the top level. But sure, I hear your point that its low-hanging fruit and Duration/Size is better to work with. Happy to add them.

For `vpcConfig`, I think we should expose a property `vpc: ec2.IVpc` similar to what we do in lambda.

**Comment by RichiCoder1** [comment]
Sorry for taking so long to circle back around on this. Gonna scope it back down to VPC (current biggest pain point IMHO) and then can do a follow up on Run Config props.

**Comment by RichiCoder1** [comment]
This is should be good to review now!

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/lib/canary.ts:410]: I think there's room to improve the customer experience here. From this [doc](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_VPC.html): 

```
If the VPC is not already enabled for internet access, you have two options:

    Enable it for internet access. For more information, see the following section [Giving internet access to your canary on a VPC](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_VPC.html#CloudWatch_Synthetics_VPC_Internet).

    If you want to keep your VPC private, you can configure the canary to send its data to CloudWatch and Amazon S3 through private VPC endpoints. If you have not already done so, you must create endpoints for these services on your VPC. For more information, see Using CloudWatch and CloudWatch Synthetics with interface VPC endpoints and [Amazon VPC Endpoints for Amazon S3](https://docs.aws.amazon.com/glue/latest/dg/vpc-endpoints-s3.html).
```

I wonder if we can check to make sure the subnet is public as a default, and provide an option like `allowPrivateSubnet` to override this check. This will force our users to reckon with canary-specific vpc considerations at synth time.

Just spitballing here, I'm not very familiar with vpcs and their use cases so I'm happy to have some dialogue about this.

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/lib/canary.ts:419]: What happens if the user provides `props.securityGroups = []`?  What do we want to happen?

**Review by kaizen3031593 — DISMISSED** [review]
Hi @RichiCoder1! Thanks for the PR - it looks like a great start. I don't have much experience with vpcs, so I'm hoping we can lean on your context to provide the best possible experience for our users.

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-synthetics/lib/canary.ts:203]: Can probably tell here where my implementation and docs take their inspiration from haha

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-synthetics/lib/canary.ts:410]: Yah, there's not a good way for us to check if the VPCe are available, so this might be a nice synth time validation. This was largely inspired by Lambda (since Synth's VPC is all just Lambda VPC), but I think this is a Synth specific caveat so worth checking.

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-synthetics/lib/canary.ts:419]: The short answer is I don't know, but if I had to hazard a guess it would attach to the default security group for a given VPC which might cause surprising behavior. Though I guess if we documented it and explicitly said `[]` == `default sg` that wouldn't be too surprising?

File: packages/aws-cdk-lib/package.json
Comment (RichiCoder1) [reply packages/aws-cdk-lib/package.json:502]: Not sure, don't know how I missed this. Might have been added by some tooling or a bad merge. I'll strip it out and see if it comes back

File: packages/@aws-cdk/aws-synthetics/test/integ.canary.ts
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-synthetics/test/integ.canary.ts:73]: Sounds good. For stack verification, do mean something like checking the underlying Lambda is in a VPC, or simulate some sort of VPC-only scenario?

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-synthetics/lib/canary.ts:419]: To clarify what happens presently with this PR, we fallback to creating an sg which allows all egress traffic.

Comment (kaizen3031593) [reply packages/@aws-cdk/aws-synthetics/lib/canary.ts:419]: Actually, I missed that you have `props.securityGroup.length > 0`. I think that is fine.

File: packages/@aws-cdk/aws-synthetics/test/integ.canary.ts
Comment (kaizen3031593) [reply packages/@aws-cdk/aws-synthetics/test/integ.canary.ts:73]: best if we can check underlying lambda is in vpc, if possible. i don't actually know how.

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-synthetics/lib/canary.ts:420]: It's a bit roundabout, we now check if all the "selected" subnets are "isolated" which is VPC verbage for no internet access (private w/ no IGW or NAT) and say how to fix it.

**Comment by RichiCoder1** [comment]
Should be cleaned up and catch things a little better!

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/lib/canary.ts:206]:
```suggestion
   * @default false
```

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/lib/canary.ts:211]: This also can't be set if `vpc` is not set, right? Needs some documentation.

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/lib/canary.ts:424]: I think we should fail synth if a user tries to set vpc related options without setting `vpc`. Right now they look to be just ignored in that case, which might be cause for confusion.

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/lib/canary.ts:431]:
```suggestion
      throw new Error("A canary in a vpc must have access to CloudWatch and S3. Your vpc must either have internet access or private vpc endpoints with 'allowPrivateSubnet=true'.\nSee https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_VPC.html");
```

File: packages/@aws-cdk/aws-synthetics/test/canary.test.ts
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/test/canary.test.ts:445]: Can you also add the following unit tests:

- use `vpcSubnets`
- no `securityGroups` means one will be created for you
- can specify a `securityGroups` correctly
- vpc with no public security groups and no `allowPrivateSubnets` fails
- vpc with no public security groups and `allowPrivateSubnets` does not fail

Please also throw all these unit tests under a 

```ts
describe('canary in a vpc', () => {
```

File: packages/@aws-cdk/aws-synthetics/test/integ.asset.expected.json
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/test/integ.asset.expected.json:1]: LOL! Thanks for deleting this

**Review by kaizen3031593 — DISMISSED** [review]
This is really coming along @RichiCoder1! Thanks for your help contributing this. A few more comments here and there, and also looking for more robust unit testing.

For full context, I'm not super familiar with vpcs, so after this round of edits I might toss this over to get this PR looked at by another CDK member.

File: packages/@aws-cdk/aws-synthetics/test/integ.asset.expected.json
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-synthetics/test/integ.asset.expected.json:1]: Np haha. When I was breaking out the other, I noticed this had no matching test :)

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-synthetics/lib/canary.ts:430]: I think the two places I took for inspiration don't explicitly do this, _but_ I think it's a good idea.

**Review by kaizen3031593 — APPROVED** [review]
Looks good to me @RichiCoder1. Approved, though I am going to wait on a second pair of eyes before we get this merged.

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (corymhall) [inline packages/@aws-cdk/aws-synthetics/lib/canary.ts:210]: I don't think this is needed. I think it should be fine to just add this to the documentation in the readme and the `vpcSubnets` docstring. Users will specify if they want to place it in isolated subnets via the `vpcSubnets` prop, asking again here feels like a "are you sure you know what you are doing?" and I don't think that is needed.

**Review by corymhall — CHANGES_REQUESTED** [review]
@RichiCoder1 just a couple of comments, the main one around implementing `IConnectable`.

Since the `Canary` class can now have security groups, it should implement `ec2.IConnectable`

```ts
export class Canary extends cdk.Resource implements ec2.IConnectable {
```

**Comment by RichiCoder1** [comment]
Nuked `allowPrivateSubnet` and added `IConnectable`! Let me know if that looks good and the implementation looks appropriate @corymhall.

**Comment by RichiCoder1** [comment]
Nuked `allowPrivateSubnet` and added `IConnectable`! Let me know if that looks good and the implementation looks appropriate @corymhall.

**Review by corymhall — CHANGES_REQUESTED** [review]
Just 1 more comment and then we are good to go!

File: packages/@aws-cdk/aws-synthetics/lib/canary.ts
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-synthetics/lib/canary.ts]: @corymhall Done! Though used `Lazy.list` as I got a warning that `listValue` was deprecated and that seemed to be the replacement.

**Review by corymhall — APPROVED** [review]

**Review by kaizen3031593 — APPROVED** [review]
Looks great @RichiCoder1. Thanks for the contribution!

**Pull request merged**
